### PR TITLE
[5.1] allow $casts to transform date and datetime fields

### PR DESCRIPTION
**Pull request by tomschlick**
This transforms fields marked as date & datetime to carbon objects. Its nice to consolidate this logic in one in certain cases instead of spreading it around to multiple variables.

### CODE REVIEW

**Comment by alexrussell** [comment]
So does this now remove the `protected $dates = [];`/`getDates()` method feature, work as well as it or clash with it?

**Comment by tomschlick** [comment]
Right now this would work very similar it but shouldn't have the same variable in both as one might try to use the other in the Carbon constructor. If deemed necessary I'll add the ability for it to work in conjunction.

My thinking is that eloquent fields like created_at, updated_at, and deleted_at would be managed by the $dates array and your own fields would be managed by the $casts. 

This came about as I was adding 20 or so fields to $casts and had to omit 5 of them as they were dates/datetimes.

**Comment by alexrussell** [comment]
I totally understand the reasoning, and I actually support the idea of removing the `$dates` feature entirely now we have `$casts`. Of course this is a major breaking change (though `$dates` could be deprecated and the casts version be made to ignore a field that's already specified in $dates or something) so I dunno how far you'd get doing that.

File: src/Illuminate/Database/Eloquent/Model.php [hunk 1]
```diff
@@ -2808,6 +2808,10 @@ protected function castAttribute($key, $value)
                 return json_decode($value, true);
             case 'collection':
                 return new BaseCollection(json_decode($value, true));
+            case 'date':
+                return Carbon::createFromFormat('Y-m-d', $value);
+            case 'datetime':
+                return Carbon::createFromFormat('Y-m-d H:i:s', $value);
```
Comment (crynobone) [inline src/Illuminate/Database/Eloquent/Model.php]: Need to get the value from format, since different database grammar would have different formatting.

Or why not use the available methods to convert it to carbon object (if possible).

Comment (tomschlick) [inline src/Illuminate/Database/Eloquent/Model.php]: You're saying that grammar can provide the format to pass into Carbon::createFromFormat()? or because the grammar is different it should just use the value and let Carbon handle it?

If its the first one can you provide a link to where grammar does that?

Comment (crynobone) [inline src/Illuminate/Database/Eloquent/Model.php]: https://github.com/laravel/framework/blob/5.1/src/Illuminate/Database/Grammar.php#L163 vs https://github.com/laravel/framework/blob/5.1/src/Illuminate/Database/Query/Grammars/SqlServerGrammar.php#L209

Comment (crynobone) [inline src/Illuminate/Database/Eloquent/Model.php]: And used in Eloquent via https://github.com/laravel/framework/blob/5.1/src/Illuminate/Database/Eloquent/Model.php#L2946

Comment (tomschlick) [inline src/Illuminate/Database/Eloquent/Model.php]: Thanks, I'll get that changed.

**Comment by tomschlick** [comment]
Actually the getDateFormat method poses some problems because it doesn't differentiate between a date and a datetime. Passing a datetime format to carbon with a date as the value throws errors. I think for this I'm just going to use Carbon::parse() and let it do the heavy lifting.

**Comment by tomschlick** [comment]
scratch that, just found the asDateTime() method. Will use that and squash these

**Comment by tomschlick** [comment]
@alexrussell with the latest changes it can work with the $dates array just fine if a property is in both.

**Comment by taylorotwell** [comment]
So in the `setAttribute` method should we have a check to see if the attribute is date castable and use `fromDateTime`. You'll see in that method we already have a check to see if it is in the `$dates` array.

**Comment by tomschlick** [comment]
@taylorotwell are you saying scan through each var whether it is defined in $casts or not as a date and check to see if it is a date formatted string? if so cast it to carbon?

I would rather see it be an explicit action with the $casts array. That way the dev has control over what is being casted and what stays as a string for legacy/performance reasons. For instance if they are just dumping it to a json output there is not need to throw it into carbon just to __toString() it back out.

Even with the current $dates array it is something you have to define per model for specific fields (other than created_at, updated_at, deleted_at).

**Comment by taylorotwell** [comment]
Hmm, I'm not sure if that's what I mean. I mean if you do like:

`$model->castedDate = Carbon::now()`

Where `castedDate` is some date field that is listed in `$casts`. It's not going to be cast back into raw string form in the internal `$attributes` array. Compare this to the behavior of the JSON cast in the `setAttribute` method. We check if the attribute being set is a JSON cast, and if so we convert it to raw string form for storage.

Basically there is an inconsistency between `$dates` and date casts when **setting** an attribute.

**Comment by tomschlick** [comment]
Ah ok, I see what you're talking about now. I'll make the necessary changes to `setAttribute()` and create an `isDateCastable()` which will check for dates/datetimes like the `isJsonCastable` checks for collections, arrays, etc

**Comment by taylorotwell** [comment]
Yep that looks pretty good.

File: src/Illuminate/Database/Eloquent/Model.php [hunk 2]
```diff
@@ -2748,6 +2748,23 @@ protected function hasCast($key)
...
+     * Determine whether a value is Date / DateTime castable for inbound manipulation.
+     *
+     * @param  string  $key
+     * @return bool
+     */
+    protected function isDateCastable($key)
+    {
+        if ($this->hasCast($key)) {
+            return in_array(
+                $this->getCastType($key), ['date', 'datetime'], true
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: Would be fine all on one line here.

Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: Just write: `return $this->hasCast($key) && in_array($this->getCastType($key), ['date', 'datetime'], true);`

Comment (tomschlick) [inline src/Illuminate/Database/Eloquent/Model.php]: was just following the format that isJsonCastable() set below it... should this change in both?

Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Model.php]: That way it's super readable because the code is clearly saying if has cast and is in array, return true, rather than some nested stuff that's harder to read and see the intent so quickly.

**Comment by GrahamCampbell** [comment]
What happens if something is already in the dates array? Will we try to double cast it? What happens then?

**Comment by tomschlick** [comment]
That is prevented here https://github.com/tomschlick/framework/blob/patch-1/src/Illuminate/Database/Eloquent/Model.php#L2918

**Comment by GrahamCampbell** [comment]
Good. :)

**Comment by tomschlick** [comment]
I'll make the line changes you mentioned @GrahamCampbell and squash

**Comment by taylorotwell** [comment]
:heart:

**Comment by jbrooksuk** [comment]
Perfect! :)
